### ISSUE TITLE
Query Log shows an extra query for some lookups

### ISSUE THREAD

**Issue by agneevX** [AdguardTeam/AdGuardHome#4386]
### Prerequisites

- [x] I am running the latest version
- [ ] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Environment

* **Version of AdGuard Home server:**
  * v0.107.5
* **How did you install AdGuard Home:**
  * 
* **How did you setup DNS configuration:**
  * 
* **If it's a router or IoT, please write device model:**
  * 
* **CPU architecture:**
  * 
* **Operating system and version:**
  * 

### Description

AdGuard shows two queries when one is made to AGH. This is definitely a bug, because the first query here took 11ms, which is not possible, because the DNS server is ~40ms away. The second query takes a more sensible 47ms.

![Screenshot (199)](https://user-images.githubusercontent.com/19761269/157899311-f27abb7f-9140-4bc4-88d4-6a23b1dd4a4b.png)
 
If I look at logs from the upstream, I see just one query made:

![Screenshot (200)](https://user-images.githubusercontent.com/19761269/157900493-beaa3679-3e25-42c4-ab09-3c75d619a142.png)

**Comment by JsBergbau** [comment]
Could it be that this is an issue in your upstream DNS server logging? It looks like upstream DNS Server has only a time resolution of seconds. If now two requests arrive within the same second it may be possible that only one of this request is recorded for that second and domain.
Still there is only 11 ms processing time whereas the server is 40 ms away, but maybe thats a different bug with showing wrong processing time. 
How is upstream DNS configured? Plain UDP 53? If so using `tcpdump port 53` could show how many requests really arrive.

**Comment by agneevX** [comment]
It's not. If two requests come in, it will get logged.

**Comment by agneevX** [comment]
> but maybe thats a different bug with showing wrong processing time.

Could be, not willing to rule that out.

**Comment by agneevX** [comment]
Bad bot

**Comment by EugeneOne1** [comment]
@agneevX, hello and apologies for the late response.

Are you able to reproduce it with one of the latest builds?  If yes, could you please collect a [verbose log](https://github.com/AdguardTeam/AdGuardHome/wiki/FAQ#verboselog) for us?  Also, if you still have this or analogous part of the query log file?  It'd be really helpful to take a look at the written data.  You may send any of above to devteam@adguard.com.

Could you please also give some information about the used environment?  Which OS/Architecture AGH runs on?  Which client performs the request?

**Comment by agneevX** [comment]
> Are you able to reproduce it with one of the latest builds?

> Could you please also give some information about the used environment? Which OS/Architecture AGH runs on? 

@EugeneOne1 I've updated the initial post with some more details.

> Which client performs the request?

This was from a standard Windows client.

While I no longer have the log, I think this is too easy to reproduce. Opening a couple of websites shows multiple lookups bearing the same domain name. Usually the first one has a lower than possible response time, while the second one shows the true (or possible) response time.

**Comment by valantur** [comment]
I'm glad to see this is being looked at. My query logs are full of examples like this one:

<img width="844" alt="duplicate query log entries" src="https://user-images.githubusercontent.com/2516392/178658922-217821ca-a1e3-4082-acbe-71aca146b996.png">

**Comment by cyb3rz3us** [comment]
@valantur --- Just an FYI, what you posted is normal behavior as those are not double log entires.  Those are entries for 2 different queries.

**Comment by valantur** [comment]
@cyb3rz3us --- Oops sorry, wrong screenshot. Here it goes:
<img width="898" alt="query log entries" src="https://user-images.githubusercontent.com/2516392/178660247-89a81bd9-a6f4-49c6-8677-4ef8761a979e.png">

I'm running Unbound on the same VM. Both entries say the request was processed by Unbound. I guess Unbound had it cached the second time around and that's why it responded in 1ms, but still AGH is sending two requests with minimal time in-between. I swear I saw worse examples, but I cleared my query log recently... oh well.

**Comment by CodeSapiens** [comment]
Same issue here
Multiple queries for the same items, all coming from the same PC.

![Screenshot 2024-03-07 231534](https://github.com/AdguardTeam/AdGuardHome/assets/40335803/6d827899-349d-4682-a2e2-9bad9fccdc7c)

**Comment by SeriousHoax** [comment]
Just want to add that even AdGuard DNS query log which is different from AdGuard Home also shows two dns queries for most requests unlike NextDNS. What is the issue?

**Comment by Cansybars** [comment]
I wanted to comment that I have the exact same issue, sometimes query may repeat more than twice, and always the cached will precede the non cached, and there may be more than one of each. switching from parallel to load balancing did not change anything either. I am running version v0.107.52, both on raspberry pi and on windows, and have the same problem with both. Overall this creates heavy load volumes and multiple redundant queries appearing for a single lookup. Hope this can be fixed.

**Comment by crafty991** [comment]
> Just want to add that even AdGuard DNS query log which is different from AdGuard Home also shows two dns queries for most requests unlike NextDNS. What is the issue?

That's right, by doing so, the free requests end immediately and it becomes unusable. NextDNS is a step forward in this.

**Comment by Sil3ntVip3r** [comment]
I traced this against the current pending-request and query-log paths.  There are two distinct cases that look identical in the UI:

1. Two client requests for the same name produce two query-log rows.  Since v0.107.61, identical cacheable requests that overlap are coalesced into one upstream exchange by default, but AdGuard Home still records one row per downstream request.  That is expected request accounting.
2. One client packet produces two query-log rows.  That would be a query-log defect.

The screenshots alone cannot distinguish these, and an upstream log with one-second granularity cannot prove how many packets arrived.  The shortest decisive reproduction would capture the same event simultaneously on the client-facing and upstream-facing interfaces, with DNS IDs and timestamps, while verbose logging is enabled and both cache and pending_requests.enabled remain on.

Useful evidence:

- one inbound packet plus two rows => logging bug;
- two inbound packets plus two rows and one upstream packet => expected coalescing/accounting;
- two inbound packets and two upstream packets => coalescing miss requiring the request flags, protocol, and timing from the capture.

A small pcap filtered to the affected client/server, the matching verbose-log interval, and the two raw query-log records would be enough to build a deterministic regression.  Client IPs and unrelated domains can be redacted.  I do not recommend changing query-log or pending-request code without that distinction.
